### DISCUSSION TITLE
Copying/templating widgets?

### DISCUSSION THREAD

**Discussion by eichin** [beeware/toga#2531] [Q&A] [answered]
I'm using `OptionContainer` to turn a one-screen interface into one that has three _slightly_ different versions of that screen - they're so similar that for the first draft I was just going to iterate over the children for one tab's `Box` and `add()` them to the next tab's `Box`.  Looking at the `Widget.add` docs that isn't going to work - `add` (sensibly) reparents them.  `copy.copy` and `copy.deepcopy` very specifically don't work for this.

Is there a reasonable "clone this widget hierarchy" mechanism? (I should probably just factor out the current widget construction code into a function that handles all three versions and builds them from scratch, it's just early enough in the prototyping that all of the code is still in `startup()` :-)

**Comment by freakboy3742** [comment, marked as answer]
There's no builtin-mechanism to clone a widget hierarchy at present; but I'm not 100% sure we need one either. As you've noted, the best approach here is likely to refactor the construction code into a function. 

A clone mechanism is likely technically possible, but it would require every widget to gain a method to clone its current state. We'd also need to make decisions about whether *some* properties wouldn't be cloned - for example, should event handlers be cloned? On the one hand, it would be possible to write event handlers that response to where they are in the layout, but the easiest approach is to use a separate handler for every widget. Lastly, we'd need to encourage users to use the "lookup by ID" mechanism as the primary way of interacting with widgets - which, while it works, is a lot less efficient that retaining a reference to the widgets that you require. 

So - on balance, I'd suggest a clone mechanism is a complication that won't improve usability that much, and a "factory" method that produces the "similar, but slightly different" heirarchy is the approach we should encourage.
